foundations/src/telemetry/metrics/gauge.rs

```diff
@@ -0,0 +1,170 @@
+use prometheus_client::encoding::text::{EncodeMetric, Encoder};
+use prometheus_client::metrics::gauge::Gauge;
+use prometheus_client::metrics::{MetricType, TypedMetric};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+/// Prometheus metric based on a gauge, but additionally records the minimum and maximum values of
+/// that gauge since the last recorded value was taken.
+///
+/// This allows a user of the metric to see the full range of values within a smaller timespan with
+/// greater precision and less overhead than a histogram. If the details of the intermediate values
+/// are required, the histogram remains a more appropriate choice.
+#[derive(Debug, Clone, Default)]
+pub struct RangeGauge {
+    gauge: Gauge<u64, AtomicU64>,
+    min: Arc<AtomicU64>,
+    max: Arc<AtomicU64>,
+}
+
+impl RangeGauge {
+    fn update_max(&self, new_max: u64) {
+        self.max.fetch_max(new_max, Ordering::AcqRel);
+    }
+
+    fn update_min(&self, new_min: u64) {
+        self.min.fetch_min(new_min, Ordering::AcqRel);
+    }
+
+    /// Increase the [`RangeGauge`] by 1, returning the previous value.
+    pub fn inc(&self) -> u64 {
+        self.inc_by(1)
+    }
+
+    /// Increase the [`RangeGauge`] by `v`, returning the previous value.
+    pub fn inc_by(&self, v: u64) -> u64 {
+        let prev = self.gauge.inc_by(v);
+        self.update_max(prev + v);
+        prev
+    }
+
+    /// Decrease the [`RangeGauge`] by 1, returning the previous value.
+    pub fn dec(&self) -> u64 {
+        self.dec_by(1)
+    }
+
+    /// Decrease the [`RangeGauge`] by `v`, returning the previous value.
+    pub fn dec_by(&self, v: u64) -> u64 {
+        let prev = self.gauge.dec_by(v);
+        self.update_min(prev - v);
+        prev
+    }
+
+    /// Sets the [`RangeGauge`] to `v`, returning the previous value.
+    pub fn set(&self, v: u64) -> u64 {
+        let prev = self.gauge.set(v);
+        self.update_max(v);
+        self.update_min(v);
+        prev
+    }
+
+    /// Get the current value of the [`RangeGauge`].
+    pub fn get(&self) -> u64 {
+        self.gauge.get()
+    }
+
+    /// Exposes the inner atomic type of the [`RangeGauge`].
+    ///
+    /// This should only be used for advanced use-cases which are not directly
+    /// supported by the library.
+    pub fn inner(&self) -> &AtomicU64 {
+        self.gauge.inner()
+    }
+}
+
+impl TypedMetric for RangeGauge {
+    const TYPE: MetricType = MetricType::Gauge;
+}
+
+impl EncodeMetric for RangeGauge {
+    fn encode(&self, mut encoder: Encoder) -> Result<(), std::io::Error> {
+        let current = self.get();
+        // Getting the current values of the metric resets the min/max
+        let min = self.min.swap(current, Ordering::Relaxed);
+        let max = self.max.swap(current, Ordering::Relaxed);
```

Comment:
What always bothered me acquiring values of multiple atomics is that it's always data race-prone, so you end up having non-atomic values under RwLock. Though, metrics could be best effort in that regard. But can you think of any scenarios where things can get wrong when reported counters are out of sync with actual values?

Reply:
The metric expects the obvious invariant that min<=current<=max.
Swapping the min/max with a potentially stale version of current could break that invariant, so the old version was not good enough.

I really want to avoid wrapping the whole metric in a mutex, in particular I don't want inc/dec locked. But locking read/reset is OK as there should only be one metric scrape at a time. The invariant only matters when you scrape it, so we can 'fix up' the values with atomic ops after reading them and attempting to reset.